cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java

```diff
@@ -573,8 +576,36 @@ public void destroy() {
       }
     }
 
-    LineageOperationsProcessor processor = new LineageOperationsProcessor(spec.getConnections(), allStageOperations,
+    if (allStageOperations.isEmpty()) {
+      return;
+    }
+
+    Map<String, Set<String>> stageInputFields = new HashMap<>();
+    Map<String, Set<String>> stageOutputFields = new HashMap<>();
+
+    for (Map.Entry<String, StageSpec> stageSpecEntry : stageSpecs.entrySet()) {
+      Set<String> inputFields = new HashSet<>();
+      Set<String> outputFields = new HashSet<>();
+      Schema outputSchema = stageSpecEntry.getValue().getOutputSchema();
+      if (outputSchema != null && outputSchema.getFields() != null) {
+        outputFields.addAll(outputSchema.getFields().stream().map(Schema.Field::getName).collect(Collectors.toSet()));
+      }
+      stageOutputFields.put(stageSpecEntry.getKey(), outputFields);
+
+      Map<String, Schema> inputSchemas = stageSpecEntry.getValue().getInputSchemas();
+      for (Map.Entry<String, Schema> entry : inputSchemas.entrySet()) {
+        if (entry.getValue().getFields() != null) {
+          inputFields.addAll(entry.getValue().getFields().stream().map(Schema.Field::getName)
+                               .collect(Collectors.toSet()));
+        }
+      }
+      stageInputFields.put(stageSpecEntry.getKey(), inputFields);
+    }
+
+    LineageOperationsProcessor processor = new LineageOperationsProcessor(spec.getConnections(), stageInputFields,
```

Comment:
if it is invalid, it will result in an error message and a failure? Or do we ignore exceptions in destroy() and still consider the run completed?

Reply:
yes. lineage recording will fail but run is still successful.

Reply:
it's probably still better to catch the exception and log a warning about being unable to write field lineage. I believe that will result in a nicer warning, and it also protects us from adding code below this block and thinking it will always run.